huaweicloud/resource_huaweicloud_bms_instance.go

```diff
@@ -0,0 +1,589 @@
+}
+
+func resourceBmsInstanceDelete(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*config.Config)
+	bmsClient, err := config.BmsV1Client(GetRegion(d, config))
+	if err != nil {
+		return fmt.Errorf("Error creating HuaweiCloud compute client: %s", err)
+	}
+	serverID := d.Id()
+	publicIP := d.Get("public_ip").(string)
+
+	resourceIDs := make([]string, 0, 2)
+	if serverID != "" {
+		resourceIDs = append(resourceIDs, serverID)
+	}
+
+	// unsubscribe the eip if necessary
+	if _, ok := d.GetOk("iptype"); ok && publicIP != "" {
+		eipClient, err := config.NetworkingV1Client(GetRegion(d, config))
+		if err != nil {
+			return fmt.Errorf("Error creating networking client: %s", err)
+		}
+
+		if eipID, err := getEipIDbyAddress(eipClient, publicIP); err == nil {
+			resourceIDs = append(resourceIDs, eipID)
+		} else {
+			log.Printf("[WARN] Error fetching EIP ID of BMS server (%s): %s", d.Id(), err)
+		}
+	}
+
+	if len(resourceIDs) > 0 {
+		if err := UnsubscribePrePaidResource(d, config, resourceIDs); err != nil {
+			return fmt.Errorf("Error unsubscribing HuaweiCloud BMS server: %s", err)
+		}
+	}
+
+	stateConf := &resource.StateChangeConf{
+		Pending:      []string{"Deleting", "ACTIVE", "SHUTOFF"},
+		Target:       []string{"DELETED"},
+		Refresh:      waitForBmsInstanceDelete(bmsClient, d.Id()),
+		Timeout:      d.Timeout(schema.TimeoutDelete),
+		Delay:        60 * time.Second,
+		PollInterval: 20 * time.Second,
+	}
+
+	_, err = stateConf.WaitForState()
+	if err != nil {
+		return fmt.Errorf("Error deleting HuaweiCloud BMS instance: %s", err)
+	}
+
+	d.SetId("")
+	return nil
+}
+func resourceBmsInstanceNicsV1(d *schema.ResourceData) []baremetalservers.Nic {
+	var nicRequests []baremetalservers.Nic
+
+	nics := d.Get("nics").([]interface{})
+	for i := range nics {
+		nic := nics[i].(map[string]interface{})
+		nicRequest := baremetalservers.Nic{
+			SubnetId:  nic["subnet_id"].(string),
+			IpAddress: nic["ip_address"].(string),
+		}
+
+		nicRequests = append(nicRequests, nicRequest)
+	}
+	return nicRequests
+}
+
+func resourceBmsInstanceDataVolumesV1(d *schema.ResourceData) []baremetalservers.DataVolume {
+	var volRequests []baremetalservers.DataVolume
+
+	vols := d.Get("data_disks").([]interface{})
+	for i := range vols {
+		vol := vols[i].(map[string]interface{})
+		volRequest := baremetalservers.DataVolume{
+			VolumeType: vol["type"].(string),
+			Size:       vol["size"].(int),
+		}
+		volRequests = append(volRequests, volRequest)
+	}
+	return volRequests
+}
+
+func resourceBmsInstanceSecGroupsV1(d *schema.ResourceData) []baremetalservers.SecurityGroup {
+	rawSecGroups := d.Get("security_groups").(*schema.Set).List()
+	secgroups := make([]baremetalservers.SecurityGroup, len(rawSecGroups))
+	for i, raw := range rawSecGroups {
+		secgroups[i] = baremetalservers.SecurityGroup{
+			ID: raw.(string),
+		}
+	}
+	return secgroups
+}
+
+func flattenBmsInstanceNicsV1(d *schema.ResourceData, meta interface{},
```

Comment:
can we use the similar function in `huaweicloud_compute_instance`?

Reply:
the type of param `address` is different